lms/djangoapps/ccx/overrides.py

```diff
@@ -130,23 +137,33 @@ def override_field_for_ccx(ccx, block, name, value):
     field = block.fields[name]
     value_json = field.to_json(value)
     serialized_value = json.dumps(value_json)
-    try:
-        override = CcxFieldOverride.objects.create(
-            ccx=ccx,
-            location=block.location,
-            field=name,
-            value=serialized_value
-        )
-    except IntegrityError:
-        transaction.commit()
-        override = CcxFieldOverride.objects.get(
-            ccx=ccx,
-            location=block.location,
-            field=name
-        )
+    has_changes = False
+
+    override = get_override_for_ccx(ccx, block, name + "_instance")
+    if override:
+        has_changes = serialized_value != override.value
+
+    if not override:
+        try:
+            override = CcxFieldOverride.objects.create(
+                ccx=ccx,
+                location=block.location,
+                field=name,
+                value=serialized_value
+            )
+            _get_overrides_for_ccx(ccx).setdefault(block.location, {})[name + "_id"] = override.id
+        except IntegrityError:
+            transaction.commit()
```

Comment:
@pdpinch, @ormsbee, here's my read on what L156 above is for:

Back when we first wrote this, we discovered a race condition that occurred because of the use of the Django ORM's `get_or_create` method.  I believe that this race condition is actually a [known issue](https://code.djangoproject.com/ticket/13906) that bites pretty much all ORMs using MySQL under multi-threaded conditions and implementing this type of insert-or-update functionality (the same bug report is in the code base for Rails as well).

It appears that the approach we took to dealing with this race condition was to attempt first to `create` a new CCX record, and in the condition that that fails due to an integrity error, update the existing record we now know must be there.  I think the purpose of the `transaction.commit()` on L156 is to reset the condition of the transaction which at that point must be in an error state.  The same could be done with `transaction.rollback()`, but that would risk losing other work done under the same transaction.  

I have written other code for another project that faced this same problem using a while loop to re-try the get_or_create call until it succeeded.  That approach has the advantage of only ever looping if the race condition is hit, where this will end up raising an error, catching it and then committing and starting a new transaction every time an existing CCX is updated.

We can't try to get an existing CCX first though, as there is no way to guarantee that between the time that no matching CCX is found and a new one is created on one thread, some other thread might have already committed a new CCX matching ours, and cause an integrity error when our new CCX is created.  

I am fully open to any suggestions of better ways to fix the `get_or_create` race condition.